pages/en/zkapps/how-to-test-a-zkapp.mdx

```diff
@@ -37,8 +37,97 @@ Because we are using Jest, it's helpful to break down all functionality of your
 
 For examples of existing tests, we recommend creating a template example using the Mina zkApp CLI via `zk project <name>` and examining the test file there. [You will see a basic example of a few tests](https://github.com/o1-labs/zkapp-cli/blob/main/templates/project-ts/src/Add.test.ts) that deploy and update the state on a smart contract using Jest.
 
+#### Creating a local blockchain
+
+The `Mina.LocalBlockchain()` method specifies a mock Mina ledger of accounts that runs locally. It contains logic for updating the ledger that can be used to test your smart contract locally.
+
+```ts
+let Local = Mina.LocalBlockchain();
+Mina.setActiveInstance(Local);
+```
+
+#### Deploying a contract locally
+
+The mock Mina local blockchain contains test accounts that can be used to deploy the contract and pay transaction fees.
+
+```ts
+// test accounts that pays all the fees, and puts additional funds into the zkapp.
+let feePayer = Local.testAccounts[0].privateKey;
+```
+
+We generate a zkApp account and a new instance of the smart contract to deploy locally for testing.
+
+```ts
+// zkapp account
+let zkAppPrivateKey = PrivateKey.random();
+let zkAppAddress = zkAppPrivateKey.toPublicKey();
+let zkAppInstance = new Add(zkAppAddress);
+```
+
+Use the test account and zkApp keys to construct a transaction to pay account creation fees and deploy your smart contract. This transaction is sent to the local blockchain. 
+
+```ts
+let txn = await Mina.transaction(feePayer, () => {
+  Party.fundNewAccount(feePayer);
+  zkAppInstance.deploy({ zkappKey: zkAppPrivateKey});
+});
+txn.send();
+```
+
+#### Writing integration tests
+
+While Jest can be used to write both unit and integration tests, It is reccomended to create linear scripts for smart contract integration tests. The scripting worflow is easier to reason about when creating large interconnected tests. It also makes debugging easier and provides good error tracing. 
+
+ A well written script tests the flow of your smart contracts expected behavior, and verifies correctness of state updates. After you test the expected behavior, it's benifical to verify the preconditions of your methods and test edge cases of your smart contract. Below is an example of a basic integration test script.
```

Comment:
Is preconditions hyphenated? No, it's not: https://en.wikipedia.org/wiki/Precondition. TIL.